README.md
![GitHub issues](https://img.shields.io/github/issues/CICCIOSGAMINO/snack-bar)
[![npm version](https://badgen.net/npm/v/@cicciosgamino/progress-ring)](https://www.npmjs.com/package/@cicciosgamino/snack-bar)
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/cicciosgamino/snack-bar)

# 🌰 \<snack-bar\>

Simple Snackbar CustomElement 🌰!

<p align="center">
  <a href="#examples">examples</a> •
  <a href="#usage">usage</a> •
  <a href="#api">api</a> •
  <a href="#accessibility">accessibility</a> •
  <a href="#todo">TODO</a> •
</p>

## Examples

![Example Snackbar](https://raw.githubusercontent.com/CICCIOSGAMINO/cicciosgamino.github.io/master/images/exampleSnackbar.gif)

```html
<snack-bar timing="5000" title="Offline"></snack-bar>
```

## 🚀 Usage

1. Install package
```bash
npm install --save @cicciosgamino/snack-bar
```

2. Import
```html
<!-- Import Js Module -->
<script type="module">
  // Importing this module registers <progress-ring> as an element that you
  // can use in this page.
  //
  // Note this import is a bare module specifier, so it must be converted
  // to a path using a server such as @web/dev-server.
  import '@cicciosgamino/snack-bar'
</script>
```

3. Place in your HTML
```html
<snack-bar timing="5000" title="Offline"></snack-bar>
```

## 🐝 API

### 📒 Properties/Attributes

| Name | Type | Default | Description
| ------- | ---------- | ------------ | --------------
| timing  | Number  | `3000`  | Snackbar active timing
| title   | String  | `🌰 ... SnackBar Title` | Snackbar title
| active  | Boolean | `` | Snackbar active attribute (present/NOT)
| closebtn| Boolean | `` | Show the close button when attribute is present


### Methods

| Name     | Description
| -------- | -------------
| `closeSnackbar() => void`  | Close the snackbar


### Events
*None*

### 🧁 CSS Custom Properties

| Name | Default | Description
| ------------------- | --------- | ----------------------------------
| `--snak-bk-color`   |   `#333`  | Snackbar background-color
| `--snack-txt-color` | `#f5f5f5` | Snackbar text color
| `--snack-font-size` | `2.1rem`  | Snackbar text / space / proportion
| `--snack-bottom`    | `0`       | Snackbar bottom
| `--snack-radius`    | `1px`     | Snackbar border-radius

### 🤖 Write HTML and JavaScript
Import the component's JavaScript module, use the component in your HTML, and control it with JavaScript, just like you would with a built-in element such as `<button>`:

```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>My Example App</title>

    <!-- Add support for Web Components to older browsers. -->
    <script src="./node_modules/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>

  </head>
  <body>
    <!-- Use Web Components in your HTML like regular built-in elements. -->
    <snack-bar url=""></snack-bar>

    <!-- The Material Web Components use standard JavaScript modules. -->
    <script type="module">

      // Importing this module registers <progress-ring> as an element that you
      // can use in this page.
      //
      // Note this import is a bare module specifier, so it must be converted
      // to a path using a server such as @web/dev-server or @vitejs.
      import '@cicciosgamino/snack-bar'

      // Standard DOM APIs work with Web Components just like they do for
      // built-in elements.
      window.addEventListener('DOMContentLoaded', () => {
        const snack = document.querySelector('snack-bar')
        setTimeout(() => {
          // active the snack
          snack.setAttribute('active', '')
        }, 4000)
        setTimeout(() => {
          // close the snack (before timing ended)
          snack.closeSnackbar()
        }, 6000)
      })
    </script>
  </body>
</html>
```

### 🚀 Serve
Serve your HTML with any server or build process that supports bare module specifier resolution (see next section) in this project i'll use vite. So check the **vite.config.js** and the **package.json** for the scripts.

vite.config.js
```javascript
import { defineConfig } from 'vite'
import { resolve } from 'path'

// https://vitejs.dev/config/
export default defineConfig({
  build: {
    lib: {
      entry: 'snack-bar.js',
      formats: ['es']
    },
    rollupOptions: {
      input: {
        main: resolve(__dirname, 'index.html')
      }
    }
  }
})
```

package.json
```json
"scripts": {
  "dev": "vite --host",
  "build": "vite build",
  "preview": "vite preview --port 3000 --host"
},
```

## Contributing

Got **something interesting** you'd like to **share**? Learn about [contributing](https://github.com/CICCIOSGAMINO/init/blob/master/CONTRIBUTING.md).

# Accessibility
TODO

# 🔧 TODO 
- [ ] Basic Unit testing

## CSS encapsulation
The elements HTML structure is initialized in a [Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM), so it is impossible to apply CSS to it. If you need to change the element's default style for any reason, you should open up a new issue (or a pull request!), describing your use case, and we'll work with you on solving the problem.

## Browser support

Browsers without native [custom element support][support] require a [polyfill][]. Legacy browsers require various other polyfills. See [`examples/index.html`][example] for details.

- Chrome
- Firefox
- Safari
- Microsoft Edge

[support]: https://caniuse.com/#feat=custom-elementsv1
[polyfill]: https://github.com/webcomponents/custom-elements


## Author

| [![@cicciosgamino](https://raw.githubusercontent.com/CICCIOSGAMINO/cicciosgamino.github.io/master/images/justme%40412x412_round.png)](https://linkedin.com/in/) 	|
|:------------------------------------------------------------------------------------------:	|
|                                    **@cicciosgamino**                                      	|

## Support
Reach out to me at one of the following places:

- [Github](https://github.com/CICCIOSGAMINO)
- [Twitter](https://twitter.com/cicciosgamino)

## Donate

Donate help and contibutions!

# License
[GNU General Public License v3.0](https://github.com/CICCIOSGAMINO/init/blob/master/LICENSE)

Made 🧑‍💻 by [@cicciosgamino](https://cicciosgamino.web.app)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3.0] - 06-06-2023

### Changed
  - Minor fix
  - Click to close the component

## [1.1.0] - 29-07-2021

### Changed
  - Change module part in package.json, link to source snack-bar.js

## [1.0.0] - 07-07-2021

### Added
  - Init the Element  [@cicciosgamino](https://github.com/CICCIOSGAMINO).

snack-bar.js
import { LitElement, html, css } from 'lit'

class SnackBar extends LitElement {
  static get styles () {
    return css`
      * {
        /*
        --snack-bk-color: #333;
        --snack-txt-color: #f5f5f5;
        --snack-font-size: 2.1rem;
        --snack-bottom: 0;
        --snack-radius: 1px; */

        /* base rem unit (10px) */
        font-size: 62.5%;

        margin: 0;
        padding: 0;
        box-sizing: border-box;
      }

      :host {
        display: none;
        animation-name: fadeOut;
        animation-duration: 0.5s;
        animation-fill-mode: forwards;
      }

      :host-context([active]) {
        /* Handle CSS when host active */
        animation-name: fadeIn;
        animation-duration: 0.5s;
        animation-fill-mode: forwards;
      }

      #snackbar {
        border-radius: var(--snack-radius, 1px);
        min-width: 100%;
        background-color: var(--snack-bk-color, #333);
        color: var(--snack-txt-color, #f5f5f5);

        text-align: center;

        position: fixed;
        left: 50%;
        transform: translate(-50%, 0);
        bottom: var(--snack-bottom, 0);
      }

      #container {
        margin: calc(var(--snack-font-size, 2.1rem) * 1.1);
      }

      p {
        font-size: var(--snack-font-size, 2.1rem);
        font-weight: 400;
        line-height: var(--snack-font-size, 2.1rem);
      }

      #snackBtn {
        width: var(--snack-font-size, 2.1rem);
        height: var(--snack-font-size, 2.1rem);

        background: none;
        border: none;
        cursor: pointer;

        position: absolute;
        top: calc(var(--snack-font-size, 2.1rem) * 1.3);
        right: 3rem;
      }

      button::after {
        content: '';
        position: absolute;
        top: -50%;
        left: -50%;
        right: 0;
        bottom: 0;
        width: calc(var(--snack-font-size, 2.1rem) * 2);
        height: calc(var(--snack-font-size, 2.1rem) * 2);
        background: #555;
        border-radius: 50%;
        z-index: -1;
        transform: scale(0);
        transition: 0.3s cubic-bezier(.95, .32, .37, 1.21);
      }

      button:hover::after {
        transform: scale(1);
      }

      button svg {
        width: var(--snack-font-size, 2.1rem);
        height: var(--snack-font-size, 2.1rem);
        fill: var(--snack-txt-color, #f5f5f5);
      }

      button:hover svg {
        transform: scale(0.95);
      }

      @media only screen and (min-width: 649px) {
        #snackbar {
          min-width: 50%;
        }
      }
 
      /* Snackbar animation */
      @keyframes fadeIn {
        from { opacity: 0; }
        to { opacity: 1; }
      }
      @keyframes fadeOut {
        from { opacity: 1; }
        to { opacity: 0; }
      }
    `
  }

  #timer
  static get properties () {
    return {
      title: {
        type: String,
        reflect: true
      },
      active: Boolean,
      timing: Number,
      closebtn: {
        type: Boolean,
        reflect: true
      }
    }
  }

  constructor () {
    super()
    this.title = '🌰 ... SnackBar Title' || ''
    this.timing = 5000
  }

  connectedCallback () {
    super.connectedCallback()
    this.addEventListener('click', this.#deactivate)

    console.log('@CLOSE-BTN >> ', this.closebtn)
  }

  disconnectedCallback () {
    clearInterval(this.#timer)
    super.disconnectedCallback()
  }

  updated (changed) {
    if (changed.has('active')) {
      if (this.hasAttribute('active')) {
        this.#activate()
      }
    }
    super.updated()
  }

  #activate () {
    this.style.display = 'block'
    clearInterval(this.#timer)
    this.#timer = setTimeout(() => {
      this.#deactivate()
    }, this.timing)
  }

  #deactivate () {
    this.removeAttribute('active')
    // listen for the fadeOut animation to end > display none
    this.addEventListener('animationend', (event) => {
      if (event.animationName === 'fadeOut') {
        this.style.display = 'none'
      }
    })
  }

  // method close the snackbar
  closeSnackbar () {
    this.#deactivate()
  }

  render () {
    return html`
      <!-- The snackbar -->
      <div id="snackbar">

        <div id="container">

          <p>
            ${this.title}
          </p>

          ${this.closebtn ?
            html`
              <!-- deprecated -->
              <button
                id="snackBtn"
                @click=${this.closeSnackbar}>
                <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 0 24 24" width="24px" fill="#000000">
                  <path d="M0 0h24v24H0V0z" fill="none"/>
                  <path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/>
                </svg>
              </button>` : null
          }

        </div>
      </div>
    `
  }
}

customElements.define('snack-bar', SnackBar)
